Data processing on a big dataset with Databricks Spark

Setting up Azure blob storage and databricks

In [ ]:
storage_account_name='bigdataassignment'

In [ ]:
storage_account_access_key = "secret key here"

In [ ]:
blob_container_name = "bigdatassignment"

In [ ]:
dbutils.fs.mount(
  source = f'wasbs://{blob_container_name}@{storage_account_name}.blob.core.windows.net',
  mount_point = f'/mnt/{blob_container_name}/',
  extra_configs = {'fs.azure.account.key.' + storage_account_name + '.blob.core.windows.net': storage_account_access_key}
)

In [ ]:
dbutils.fs.ls("/mnt/bigdatassignment/")

Out[2]: [FileInfo(path='dbfs:/mnt/bigdatassignment/green_taxi_2015.parquet', name='green_taxi_2015.parquet', size=404556105, modificationTime=1727786438000),
 FileInfo(path='dbfs:/mnt/bigdatassignment/green_taxi_2015_csv/', name='green_taxi_2015_csv/', size=0, modificationTime=1728124366000),
 FileInfo(path='dbfs:/mnt/bigdatassignment/green_taxi_2016.parquet', name='green_taxi_2016.parquet', size=346731598, modificationTime=1727786342000),
 FileInfo(path='dbfs:/mnt/bigdatassignment/green_taxi_2017.parquet', name='green_taxi_2017.parquet', size=251504885, modificationTime=1727786161000),
 FileInfo(path='dbfs:/mnt/bigdatassignment/green_taxi_2018.parquet', name='green_taxi_2018.parquet', size=193874396, modificationTime=1727786055000),
 FileInfo(path='dbfs:/mnt/bigdatassignment/green_taxi_2019.parquet', name='green_taxi_2019.parquet', size=142163406, modificationTime=1727785966000),
 FileInfo(path='dbfs:/mnt/bigdatassignment/green_taxi_2020.parquet', name='green_taxi_2020.parquet', size=

In [ ]:
# Load green taxi parquet files
green_taxi_2015 = spark.read.parquet("/mnt/bigdatassignment/green_taxi_2015.parquet")
green_taxi_2016 = spark.read.parquet("/mnt/bigdatassignment/green_taxi_2016.parquet")
green_taxi_2017 = spark.read.parquet("/mnt/bigdatassignment/green_taxi_2017.parquet")
green_taxi_2018 = spark.read.parquet("/mnt/bigdatassignment/green_taxi_2018.parquet")
green_taxi_2019 = spark.read.parquet("/mnt/bigdatassignment/green_taxi_2019.parquet")
green_taxi_2020 = spark.read.parquet("/mnt/bigdatassignment/green_taxi_2020.parquet")
green_taxi_2021 = spark.read.parquet("/mnt/bigdatassignment/green_taxi_2021.parquet")
green_taxi_2022 = spark.read.parquet("/mnt/bigdatassignment/green_taxi_2022.parquet")

# Load yellow taxi parquet files
yellow_taxi_2015 = spark.read.parquet("/mnt/bigdatassignment/yellow_taxi_2015.parquet")
yellow_taxi_2016 = spark.read.parquet("/mnt/bigdatassignment/yellow_taxi_2016.parquet")
yellow_taxi_2017 = spark.read.parquet("/mnt/bigdatassignment/yellow_taxi_2017.parquet")
yellow_taxi_2018 = spark.read.parquet("/mnt/bigdatassignment/yellow_taxi_2018.parquet")
yellow_taxi_2019 = spark.read.parquet("/mnt/bigdatassignment/yellow_taxi_2019.parquet")
yellow_taxi_2020 = spark.read.parquet("/mnt/bigdatassignment/yellow_taxi_2020.parquet")
yellow_taxi_2021 = spark.read.parquet("/mnt/bigdatassignment/yellow_taxi_2021.parquet")
yellow_taxi_2022 = spark.read.parquet("/mnt/bigdatassignment/yellow_taxi_2022.parquet")


Writing files to DBFS

In [ ]:
# Save green taxi parquet files to DBFS
green_taxi_2015.write.format("parquet").save("/dbfs/tmp/green_taxi_2015.parquet")
green_taxi_2016.write.format("parquet").save("/dbfs/tmp/green_taxi_2016.parquet")
green_taxi_2017.write.format("parquet").save("/dbfs/tmp/green_taxi_2017.parquet")
green_taxi_2018.write.format("parquet").save("/dbfs/tmp/green_taxi_2018.parquet")
green_taxi_2019.write.format("parquet").save("/dbfs/tmp/green_taxi_2019.parquet")
green_taxi_2020.write.format("parquet").save("/dbfs/tmp/green_taxi_2020.parquet")
green_taxi_2021.write.format("parquet").save("/dbfs/tmp/green_taxi_2021.parquet")
green_taxi_2022.write.format("parquet").save("/dbfs/tmp/green_taxi_2022.parquet")

# Save yellow taxi parquet files to DBFS
yellow_taxi_2015.write.format("parquet").save("/dbfs/tmp/yellow_taxi_2015.parquet")
yellow_taxi_2016.write.format("parquet").save("/dbfs/tmp/yellow_taxi_2016.parquet")
yellow_taxi_2017.write.format("parquet").save("/dbfs/tmp/yellow_taxi_2017.parquet")
yellow_taxi_2018.write.format("parquet").save("/dbfs/tmp/yellow_taxi_2018.parquet")
yellow_taxi_2019.write.format("parquet").save("/dbfs/tmp/yellow_taxi_2019.parquet")
yellow_taxi_2020.write.format("parquet").save("/dbfs/tmp/yellow_taxi_2020.parquet")
yellow_taxi_2021.write.format("parquet").save("/dbfs/tmp/yellow_taxi_2021.parquet")
yellow_taxi_2022.write.format("parquet").save("/dbfs/tmp/yellow_taxi_2022.parquet")


In [ ]:
# Verify files in DBFS
dbutils.fs.ls("/dbfs/tmp/")


[FileInfo(path='dbfs:/dbfs/tmp/combined_cleaned_taxi_data.parquet/', name='combined_cleaned_taxi_data.parquet/', size=0, modificationTime=0),
 FileInfo(path='dbfs:/dbfs/tmp/combined_taxi_data_with_location.parquet/', name='combined_taxi_data_with_location.parquet/', size=0, modificationTime=0),
 FileInfo(path='dbfs:/dbfs/tmp/green_taxi_2015.parquet/', name='green_taxi_2015.parquet/', size=0, modificationTime=0),
 FileInfo(path='dbfs:/dbfs/tmp/green_taxi_2015_csv/', name='green_taxi_2015_csv/', size=0, modificationTime=0),
 FileInfo(path='dbfs:/dbfs/tmp/green_taxi_2016.parquet/', name='green_taxi_2016.parquet/', size=0, modificationTime=0),
 FileInfo(path='dbfs:/dbfs/tmp/green_taxi_2017.parquet/', name='green_taxi_2017.parquet/', size=0, modificationTime=0),
 FileInfo(path='dbfs:/dbfs/tmp/green_taxi_2018.parquet/', name='green_taxi_2018.parquet/', size=0, modificationTime=0),
 FileInfo(path='dbfs:/dbfs/tmp/green_taxi_2019.parquet/', name='green_taxi_2019.parquet/', size=0, modificationT

Verifying data

In [ ]:
# Load all green taxi Parquet files from DBFS
green_taxi_df = spark.read.parquet("/dbfs/tmp/green_taxi_*.parquet")
green_taxi_count = green_taxi_df.count()
print(f"Total number of rows for Green Taxi: {green_taxi_count}")


Total number of rows for Green Taxi: 66200401


In [ ]:
# Load all yellow taxi Parquet files from DBFS
yellow_taxi_df = spark.read.parquet("/dbfs/tmp/yellow_taxi_*.parquet")
yellow_taxi_count = yellow_taxi_df.count()
print(f"Total number of rows for Yellow Taxi: {yellow_taxi_count}")


Total number of rows for Yellow Taxi: 663055251


In [ ]:
#load reference file
taxi_zone_lookup = spark.read.csv("/mnt/bigdatassignment/taxi_zone_lookup.csv")

In [ ]:
taxi_zone_lookup.write.format("csv").save("/dbfs/tmp/taxi_zone_lookup")

Testing to analyze the file size if the data was in CSV format instead of parquet.

In [ ]:

green_taxi_2015_df = spark.read.parquet("/dbfs/tmp/green_taxi_2015.parquet")
# Convert the Parquet file to CSV and save it to DBFS (Databricks File System)
green_taxi_2015_df.write.format("csv").save("/dbfs/tmp/green_taxi_2015_csv")


In [ ]:
# List the CSV files
display(dbutils.fs.ls("/dbfs/tmp/green_taxi_2015_csv/"))


In [ ]:
green_taxi_2015_df = spark.read.parquet("/dbfs/tmp/green_taxi_2015.parquet")
# Convert the Parquet file to CSV and save it to azure
green_taxi_2015_df.write.format("csv").save("/mnt/bigdatassignment/green_taxi_2015_csv")


In [ ]:
display(dbutils.fs.ls("/dbfs/tmp/"))

path name size modificationTime dbfs:/dbfs/tmp/green_taxi_2015.parquet/ green_taxi_2015.parquet/ 0 0 dbfs:/dbfs/tmp/green_taxi_2015_csv/ green_taxi_2015_csv/ 0 0 dbfs:/dbfs/tmp/green_taxi_2016.parquet/ green_taxi_2016.parquet/ 0 0 dbfs:/dbfs/tmp/green_taxi_2017.parquet/ green_taxi_2017.parquet/ 0 0 dbfs:/dbfs/tmp/green_taxi_2018.parquet/ green_taxi_2018.parquet/ 0 0 dbfs:/dbfs/tmp/green_taxi_2019.parquet/ green_taxi_2019.parquet/ 0 0 dbfs:/dbfs/tmp/green_taxi_2020.parquet/ green_taxi_2020.parquet/ 0 0 dbfs:/dbfs/tmp/green_taxi_2021.parquet/ green_taxi_2021.parquet/ 0 0 dbfs:/dbfs/tmp/green_taxi_2022.parquet/ green_taxi_2022.parquet/ 0 0 dbfs:/dbfs/tmp/taxi_zone_lookup/ taxi_zone_lookup/ 0 0 dbfs:/dbfs/tmp/yellow_taxi_2015.parquet/ yellow_taxi_2015.parquet/ 0 0 dbfs:/dbfs/tmp/yellow_taxi_2016.parquet/ yellow_taxi_2016.parquet/ 0 0 dbfs:/dbfs/tmp/yellow_taxi_2017.parquet/ yellow_taxi_2017.parquet/ 0 0 dbfs:/dbfs/tmp/yellow_taxi_2018.parquet/ yellow_taxi_2018.parquet/ 0 0 dbfs:/dbfs/tmp/yellow_taxi_2019.parquet/ yellow_taxi_2019.parquet/ 0 0 dbfs:/dbfs/tmp/yellow_taxi_2020.parquet/ yellow_taxi_2020.parquet/ 0 0 dbfs:/dbfs/tmp/yellow_taxi_2021.parquet/ yellow_taxi_2021.parquet/ 0 0 dbfs:/dbfs/tmp/yellow_taxi_2022.parquet/ yellow_taxi_2022.parquet/ 0 0

Data cleaning and analysis

In [ ]:
from pyspark.sql.functions import col, unix_timestamp, expr

# Load the taxi dataset 
yellow_taxi_df = spark.read.parquet("/dbfs/tmp/yellow_*.parquet")

green_taxi_df = spark.read.parquet("/dbfs/tmp/green_*.parquet")



In [ ]:
# performing necessary data transformation
from pyspark.sql.functions import col, unix_timestamp

# Define the cleaning logic as a function to apply it to both datasets
def clean_data1(df):
    # Remove trips where drop-off is before pick-up
    df = df.filter(col("lpep_dropoff_datetime") > col("lpep_pickup_datetime"))
    
    # Remove trips outside of the date range (2015-2022)
    df = df.filter((col("lpep_pickup_datetime") >= "2015-01-01") & 
                   (col("lpep_dropoff_datetime") <= "2022-12-31"))
    
    # Calculate trip duration and speed
    df = df.withColumn("duration_hours", 
                       (unix_timestamp(col("lpep_dropoff_datetime")) - unix_timestamp(col("lpep_pickup_datetime"))) / 3600)
    df = df.withColumn("speed_kmh", col("trip_distance") / col("duration_hours"))
    
    # Remove trips with negative or zero speed and very high speeds (> 100 km/h)
    df = df.filter((col("speed_kmh") > 0) & (col("speed_kmh") <= 100))
    
    # Remove trips with unrealistic duration (less than 1 minute or more than 3 hours)
    df = df.filter((col("duration_hours") > 1/60) & (col("duration_hours") <= 3))
    
    # Remove trips with unrealistic distances (less than 0.1 km or more than 100 km)
    df = df.filter((col("trip_distance") > 0.1) & (col("trip_distance") <= 100))
    
    # Remove trips with unrealistic passenger count (e.g., between 1 and 6)
    df = df.filter((col("passenger_count") > 0) & (col("passenger_count") <= 6))
    
    return df

def clean_data2(df):
    # Remove trips where drop-off is before pick-up
    df = df.filter(col("tpep_dropoff_datetime") > col("tpep_pickup_datetime"))
    
    # Remove trips outside of the date range (2015-2022)
    df = df.filter((col("tpep_pickup_datetime") >= "2015-01-01") & 
                   (col("tpep_dropoff_datetime") <= "2022-12-31"))
    
    # Calculate trip duration and speed
    df = df.withColumn("duration_hours", 
                       (unix_timestamp(col("tpep_dropoff_datetime")) - unix_timestamp(col("tpep_pickup_datetime"))) / 3600)
    df = df.withColumn("speed_kmh", col("trip_distance") / col("duration_hours"))
    
    # Remove trips with negative or zero speed and very high speeds (> 100 km/h)
    df = df.filter((col("speed_kmh") > 0) & (col("speed_kmh") <= 100))
    
    # Remove trips with unrealistic duration (less than 1 minute or more than 3 hours)
    df = df.filter((col("duration_hours") > 1/60) & (col("duration_hours") <= 3))
    
    # Remove trips with unrealistic distances (less than 0.1 km or more than 100 km)
    df = df.filter((col("trip_distance") > 0.1) & (col("trip_distance") <= 100))
    
    # Remove trips with unrealistic passenger count (e.g., between 1 and 6)
    df = df.filter((col("passenger_count") > 0) & (col("passenger_count") <= 6))
    
    return df    

# Apply the cleaning logic to both datasets
cleaned_yellow_taxi_df = clean_data2(yellow_taxi_df)
cleaned_green_taxi_df = clean_data1(green_taxi_df)


In [ ]:
# Check schema of both DataFrames
cleaned_yellow_taxi_df.printSchema()
cleaned_green_taxi_df.printSchema()


root
 |-- VendorID: long (nullable = true)
 |-- tpep_pickup_datetime: timestamp (nullable = true)
 |-- tpep_dropoff_datetime: timestamp (nullable = true)
 |-- passenger_count: double (nullable = true)
 |-- trip_distance: double (nullable = true)
 |-- RatecodeID: double (nullable = true)
 |-- store_and_fwd_flag: string (nullable = true)
 |-- PULocationID: long (nullable = true)
 |-- DOLocationID: long (nullable = true)
 |-- payment_type: long (nullable = true)
 |-- fare_amount: double (nullable = true)
 |-- extra: double (nullable = true)
 |-- mta_tax: double (nullable = true)
 |-- tip_amount: double (nullable = true)
 |-- tolls_amount: double (nullable = true)
 |-- improvement_surcharge: double (nullable = true)
 |-- total_amount: double (nullable = true)
 |-- congestion_surcharge: double (nullable = true)
 |-- airport_fee: double (nullable = true)
 |-- duration_hours: double (nullable = true)
 |-- speed_kmh: double (nullable = true)

root
 |-- VendorID: long (nullable = true)
 |-- lpe

In [ ]:
# Rename lpep_pickup_datetime and lpep_dropoff_datetime in green_taxi_df
cleaned_green_taxi_df = cleaned_green_taxi_df.withColumnRenamed("lpep_pickup_datetime", "pickup_datetime") \
                                             .withColumnRenamed("lpep_dropoff_datetime", "dropoff_datetime")

# Rename tpep_pickup_datetime and tpep_dropoff_datetime in yellow_taxi_df
cleaned_yellow_taxi_df = cleaned_yellow_taxi_df.withColumnRenamed("tpep_pickup_datetime", "pickup_datetime") \
                                               .withColumnRenamed("tpep_dropoff_datetime", "dropoff_datetime")


In [ ]:
# Combine the cleaned datasets
combined_taxi_df = cleaned_yellow_taxi_df.unionByName(cleaned_green_taxi_df, allowMissingColumns=True)


In [ ]:
# Display the schema of the combined DataFrame
combined_taxi_df.printSchema()

# Show the first few rows of the combined DataFrame
combined_taxi_df.show(5)


com.databricks.backend.common.rpc.CommandCancelledException
	at com.databricks.spark.chauffeur.ExecContextState.cancel(ExecContextState.scala:445)
	at com.databricks.spark.chauffeur.ExecutionContextManagerV1.cancelExecution(ExecutionContextManagerV1.scala:460)
	at com.databricks.spark.chauffeur.ChauffeurState.$anonfun$process$1(ChauffeurState.scala:577)
	at com.databricks.logging.UsageLogging.$anonfun$recordOperation$1(UsageLogging.scala:527)
	at com.databricks.logging.UsageLogging.executeThunkAndCaptureResultTags$1(UsageLogging.scala:631)
	at com.databricks.logging.UsageLogging.$anonfun$recordOperationWithResultTags$4(UsageLogging.scala:651)
	at com.databricks.logging.AttributionContextTracing.$anonfun$withAttributionContext$1(AttributionContextTracing.scala:48)
	at com.databricks.logging.AttributionContext$.$anonfun$withValue$1(AttributionContext.scala:276)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:62)
	at com.databricks.logging.AttributionContext$.withValue(Attr

In [ ]:
# Load the location data
location_df = spark.read.csv("/dbfs/tmp/taxi_zone_lookup", header=True, inferSchema=True)

# Show the schema of the location data
location_df.printSchema()


root
 |-- LocationID: integer (nullable = true)
 |-- Borough: string (nullable = true)
 |-- Zone: string (nullable = true)
 |-- service_zone: string (nullable = true)



In [ ]:
# Perform the join for pickup location using alias
combined_with_pickup_location_df = combined_taxi_df.alias("taxi") \
    .join(location_df.alias("pickup_location"), 
          col("taxi.PULocationID") == col("pickup_location.LocationID"), "left") \
    .select("taxi.*", 
            col("pickup_location.Borough").alias("pickup_borough"),
            col("pickup_location.Zone").alias("pickup_zone"),
            col("pickup_location.service_zone").alias("pickup_service_zone"))

# Perform the join for dropoff location using alias
combined_with_location_df = combined_with_pickup_location_df.alias("taxi") \
    .join(location_df.alias("dropoff_location"), 
          col("taxi.DOLocationID") == col("dropoff_location.LocationID"), "left") \
    .select("taxi.*", 
            col("dropoff_location.Borough").alias("dropoff_borough"),
            col("dropoff_location.Zone").alias("dropoff_zone"),
            col("dropoff_location.service_zone").alias("dropoff_service_zone"))


In [ ]:
# Verify the schema of the combined DataFrame
combined_with_location_df.printSchema()

# Show some sample rows
combined_with_location_df.show(5)


root
 |-- VendorID: long (nullable = true)
 |-- pickup_datetime: timestamp (nullable = true)
 |-- dropoff_datetime: timestamp (nullable = true)
 |-- passenger_count: double (nullable = true)
 |-- trip_distance: double (nullable = true)
 |-- RatecodeID: double (nullable = true)
 |-- store_and_fwd_flag: string (nullable = true)
 |-- PULocationID: long (nullable = true)
 |-- DOLocationID: long (nullable = true)
 |-- payment_type: double (nullable = true)
 |-- fare_amount: double (nullable = true)
 |-- extra: double (nullable = true)
 |-- mta_tax: double (nullable = true)
 |-- tip_amount: double (nullable = true)
 |-- tolls_amount: double (nullable = true)
 |-- improvement_surcharge: double (nullable = true)
 |-- total_amount: double (nullable = true)
 |-- congestion_surcharge: double (nullable = true)
 |-- airport_fee: double (nullable = true)
 |-- duration_hours: double (nullable = true)
 |-- speed_kmh: double (nullable = true)
 |-- ehail_fee: double (nullable = true)
 |-- trip_type: dou

Saving to parquet and table

In [ ]:
# Export the combined DataFrame to Parquet format
combined_with_location_df.write.format("parquet").save("/dbfs/tmp/combined.parquet")


In [ ]:
# Saving as a persistent table  
combined_with_location_df.write.saveAsTable("combined_taxi_data_tables")

# Query the table
spark.sql("SELECT * FROM combined_taxi_data_table LIMIT 10").show()


In [ ]:
combined_with_location_df.cache()

DataFrame[VendorID: bigint, pickup_datetime: timestamp, dropoff_datetime: timestamp, passenger_count: double, trip_distance: double, RatecodeID: double, store_and_fwd_flag: string, PULocationID: bigint, DOLocationID: bigint, payment_type: double, fare_amount: double, extra: double, mta_tax: double, tip_amount: double, tolls_amount: double, improvement_surcharge: double, total_amount: double, congestion_surcharge: double, airport_fee: double, duration_hours: double, speed_kmh: double, ehail_fee: double, trip_type: double, pickup_borough: string, pickup_zone: string, pickup_service_zone: string, dropoff_borough: string, dropoff_zone: string, dropoff_service_zone: string]

In [ ]:
# Create a temporary view for querying in-memory
combined_with_location_df.createOrReplaceTempView("combined_taxi_data_view")

# Query the view with SQL
#spark.sql("SELECT * FROM combined_taxi_data_view LIMIT 10").show()


PART 2 Question 1

In [ ]:
spark.sql("SELECT DATE_FORMAT(pickup_datetime, 'yyyy-MM') AS year_month, COUNT(*) AS total_trips,DATE_FORMAT(pickup_datetime, 'EEEE') AS day_with_most_trips,HOUR(pickup_datetime) AS hour_with_most_trips,ROUND(AVG(passenger_count), 2) AS avg_passenger_count,ROUND(AVG(total_amount), 2) AS avg_amount_per_trip,ROUND(AVG(total_amount / NULLIF(passenger_count, 0)), 2) AS avg_amount_per_passenger FROM combined_taxi_data_view GROUP BY DATE_FORMAT(pickup_datetime, 'yyyy-MM'),DATE_FORMAT(pickup_datetime, 'EEEE'),HOUR(pickup_datetime) ORDER BY year_month").show();


+----------+-----------+-------------------+--------------------+-------------------+-------------------+------------------------+
|year_month|total_trips|day_with_most_trips|hour_with_most_trips|avg_passenger_count|avg_amount_per_trip|avg_amount_per_passenger|
+----------+-----------+-------------------+--------------------+-------------------+-------------------+------------------------+
|   2015-01|      82319|             Monday|                  13|               1.63|              14.01|                   11.48|
|   2015-01|      45095|           Saturday|                   4|                1.7|              16.58|                   13.37|
|   2015-01|      54453|             Sunday|                  23|               1.65|              16.99|                   13.94|
|   2015-01|      97571|            Tuesday|                  21|               1.65|              14.75|                   12.11|
|   2015-01|      23715|             Sunday|                   7|               1.5

Question 2


In [ ]:
spark.sql("SELECT pickup_service_zone AS taxi_color, ROUND(AVG((unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60), 2) AS avg_duration_minutes,ROUND(percentile_approx((unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60, 0.5), 2) AS median_duration_minutes,ROUND(MIN((unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60), 2) AS min_duration_minutes,ROUND(MAX((unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60), 2) AS max_duration_minutes,ROUND(AVG(trip_distance), 2) AS avg_distance_km,ROUND(percentile_approx(trip_distance, 0.5), 2) AS median_distance_km,ROUND(MIN(trip_distance), 2) AS min_distance_km,ROUND(MAX(trip_distance) , 2) AS max_distance_km,ROUND(AVG(trip_distance / ((unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 3600)), 2) AS avg_speed_kmh,ROUND(percentile_approx(trip_distance / ((unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 3600), 0.5) , 2) AS median_speed_kmh,ROUND(MIN(trip_distance / ((unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 3600)), 2) AS min_speed_kmh,ROUND(MAX(trip_distance / ((unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 3600)), 2) AS max_speed_kmh FROM combined_taxi_data_view GROUP BY taxi_color;").show()

Question 3

In [ ]:
spark.sql("SELECT pickup_service_zone,pickup_borough,dropoff_borough,DATE_FORMAT(pickup_datetime, 'yyyy-MM') AS month,DATE_FORMAT(pickup_datetime, 'EEEE') AS day_of_week,HOUR(pickup_datetime) AS hour,COUNT(*) AS total_trips,ROUND(AVG(trip_distance), 2) AS avg_distance_km,ROUND(AVG(total_amount), 2) AS avg_amount_per_trip,ROUND(SUM(total_amount), 2) AS total_amount_paid FROM combined_taxi_data_view GROUP BY pickup_service_zone, pickup_borough, dropoff_borough, DATE_FORMAT(pickup_datetime, 'yyyy-MM'), DATE_FORMAT(pickup_datetime, 'EEEE'), HOUR(pickup_datetime)ORDER BY pickup_service_zone, pickup_borough, dropoff_borough, month, day_of_week, hour;").show()

Question 4

In [ ]:
spark.sql("SELECT (COUNT(CASE WHEN tip_amount > 0 THEN 1 END) * 100.0 / COUNT(*)) AS percentage_trips_with_tips FROM combined_taxi_data_view;").show()


Question 5

In [ ]:
spark.sql("SELECT (COUNT(CASE WHEN tip_amount >= 5 THEN 1 END) * 100.0 / COUNT(*)) AS percentage_trips_with_tips_at_least_5 FROM combined_taxi_data_view WHERE tip_amount > 0;").show()


Question 6

In [ ]:
spark.sql("SELECT CASE WHEN (unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60 < 5 THEN 'Under 5 Mins' WHEN (unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60 >= 5 AND (unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60 < 10 THEN 'From 5 mins to 10 mins' WHEN (unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60 >= 10 AND (unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60 < 20 THEN 'From 10 mins to 20 mins' WHEN (unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60 >= 20 AND (unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60 < 30 THEN 'From 20 mins to 30 mins' WHEN (unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60 >= 30 AND (unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60 < 60 THEN 'From 30 mins to 60 mins' ELSE 'At least 60 mins' END AS duration_bin, ROUND(AVG(trip_distance / ((unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 3600)) , 2) AS avg_speed_kmh, ROUND(AVG(trip_distance / total_amount), 2) AS avg_distance_per_dollar FROM combined_taxi_data_view GROUP BY CASE WHEN (unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60 < 5 THEN 'Under 5 Mins' WHEN (unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60 >= 5 AND (unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60 < 10 THEN 'From 5 mins to 10 mins' WHEN (unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60 >= 10 AND (unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60 < 20 THEN 'From 10 mins to 20 mins' WHEN (unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60 >= 20 AND (unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60 < 30 THEN 'From 20 mins to 30 mins' WHEN (unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60 >= 30 AND (unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60 < 60 THEN 'From 30 mins to 60 mins' ELSE 'At least 60 mins' END ORDER BY duration_bin;").show()


Question 7


In [ ]:
spark.sql("SELECT CASE WHEN (unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60 < 5 THEN 'Under 5 Mins' WHEN (unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60 >= 5 AND (unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60 < 10 THEN 'From 5 mins to 10 mins' WHEN (unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60 >= 10 AND (unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60 < 20 THEN 'From 10 mins to 20 mins' WHEN (unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60 >= 20 AND (unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60 < 30 THEN 'From 20 mins to 30 mins' WHEN (unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60 >= 30 AND (unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60 < 60 THEN 'From 30 mins to 60 mins' ELSE 'At least 60 mins' END AS duration_bin,ROUND(AVG(total_amount / ((unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60)), 2) AS avg_fare_per_minute,ROUND(AVG(total_amount), 2) AS avg_total_fare FROM combined_taxi_data_view GROUP BY CASE WHEN (unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60 < 5 THEN 'Under 5 Mins' WHEN (unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60 >= 5 AND (unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60 < 10 THEN 'From 5 mins to 10 mins' WHEN (unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60 >= 10 AND (unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60 < 20 THEN 'From 10 mins to 20 mins' WHEN (unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60 >= 20 AND (unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60 < 30 THEN 'From 20 mins to 30 mins' WHEN (unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60 >= 30 AND (unix_timestamp(dropoff_datetime) - unix_timestamp(pickup_datetime)) / 60 < 60 THEN 'From 30 mins to 60 mins' ELSE 'At least 60 mins' END ORDER BY avg_fare_per_minute DESC;").show()


To maximize income, taxi drivers should target longer trips, specifically those lasting at least 60 minutes. This duration bin offers the highest average distance per dollar at 0.36 km/$, indicating that these trips are the most cost-efficient. With a reasonable average speed of 14.01 km/h, these longer journeys optimize earnings by avoiding heavy traffic and reducing the frequency of short trips that typically incur higher operational costs. Additionally, longer trips tend to result in higher customer satisfaction, which can lead to better tips. Therefore, focusing on longer routes such as airport transfers or inter-city travels can be more lucrative and provide more stable earnings for drivers.

PART 3: Machine Learning

Building baseline model

In [ ]:
from pyspark.sql.functions import avg, lit
from pyspark.ml.evaluation import RegressionEvaluator

# Calculate the average total_amount for baseline model
filtered_df = combined_with_location_d.sample(False, 0.5).filter(col("total_amount").isNotNull())

baseline_avg = filtered_df.filter("pickup_datetime < '2022-10-01'") \
                          .agg(avg("total_amount").alias("baseline_avg")) \
                          .collect()[0]["baseline_avg"]

# Add the baseline prediction column
baseline_df = combined_with_location_d.withColumn("baseline_prediction", lit(baseline_avg))

# Calculate RMSE for the baseline model
evaluator = RegressionEvaluator(labelCol="total_amount", predictionCol="baseline_prediction", metricName="rmse")
baseline_rmse = evaluator.evaluate(baseline_df)

print("Baseline RMSE:", baseline_rmse)


Baseline RMSE: 78.25110258177644


Preparing data for training and testing <br>
Building two models Linear Regressor and Random Forest Regressor <br> Defining the pipelines<br>Training and evaluating each models

In [ ]:
from pyspark.ml.feature import StringIndexer, VectorAssembler
from pyspark.ml.regression import LinearRegression, RandomForestRegressor
from pyspark.ml import Pipeline
from pyspark.sql.functions import col
from pyspark.ml.evaluation import RegressionEvaluator

# Index categorical columns
pickup_indexer = StringIndexer(inputCol="pickup_borough", outputCol="pickup_borough_index")
dropoff_indexer = StringIndexer(inputCol="dropoff_borough", outputCol="dropoff_borough_index")

# Select features for model training (excluding fare_amount and tolls_amount)
selected_columns = ["passenger_count", "trip_distance", "pickup_borough", "dropoff_borough", "congestion_surcharge", "airport_fee"]

# Handle Nulls and Prepare Data
combined_with_location_df = combined_with_location_df.fillna({
    "passenger_count": 1,  # Assuming default 1 passenger if null
    "trip_distance": 0,    # Assuming 0 km if null
    "congestion_surcharge": 0,  # Assuming no surcharge if null
    "airport_fee": 0,      # Assuming no airport fee if null
    "pickup_borough": 'Unknown',  # Default category if null
    "dropoff_borough": 'Unknown'  # Default category if null
})

# Prepare the training data by filtering and sampling
train_data = combined_with_location_df.filter("pickup_datetime < '2022-10-01'") \
                                      .select(selected_columns + ["total_amount"]) \
                                      .sample(False, 0.5)  

# Prepare the test data (October, November, December 2022)
test_data = combined_with_location_df.filter("pickup_datetime >= '2022-10-01'") \
                                     .select(selected_columns + ["total_amount"]).sample(False, 0.5)

# Assemble numerical features along with the indexed categorical features into a single feature vector
assembler = VectorAssembler(
    inputCols=["passenger_count", "trip_distance", "congestion_surcharge", "airport_fee", 
               "pickup_borough_index", "dropoff_borough_index"],
    outputCol="features"
)

# Define a Linear Regression model on the 'total_amount'
lr = LinearRegression(featuresCol="features", labelCol="total_amount")

# Define a Random Forest Regressor model
rf = RandomForestRegressor(numTrees=2, maxDepth=1, featuresCol='features', labelCol='total_amount')

# Pipelines
lr_pipeline = Pipeline(stages=[pickup_indexer, dropoff_indexer, assembler, lr])
rf_pipeline = Pipeline(stages=[pickup_indexer, dropoff_indexer, assembler, rf])

# Run the pipeline for Linear Regression
lr_model = lr_pipeline.fit(train_data)


evaluator = RegressionEvaluator(labelCol="total_amount", predictionCol="prediction", metricName="rmse")
lr_predictions = lr_model.transform(test_data)
lr_rmse = evaluator.evaluate(lr_predictions)
print(f"Linear Regression RMSE: {lr_rmse}")

In [ ]:
rf_model = rf_pipeline.fit(train_data)
rf_predictions = rf_model.transform(test_data)
rf_rmse = evaluator.evaluate(rf_predictions)
print(f"Random Forest RMSE: {rf_rmse}")


Choosing the best model (The model with the lowest RMSE will be chosen as the best model)

In [ ]:
best_model = "Linear Regression" if lr_rmse < rf_rmse else "Random Forest"
best_rmse = min(lr_rmse, rf_rmse)
print("Best Model:", best_model)
print("Best Model RMSE:", best_rmse)


In the comparison of these RMSE values, the model with the lowest error was identified as the best predictor. The Linear Regression model outperformed the Random Forest model and baseline model in this case, demonstrating its effectiveness for the dataset.